c++/src/kj/miniposix.h

```diff
@@ -91,8 +92,24 @@ inline int close(int fd) {
 // We're on a POSIX system or MinGW which already defines the wrappers for us.
 
 using ::ssize_t;
-using ::read;
-using ::write;
+
+#if __APPLE__
+// Workaround for FB8934446 - even though the APIs take size_t arguments they're (incorrectly)
+// internally limited to INT_MAX.
+#define KJ_OS_MAX_IO INT_MAX
+#elif INTPTR_MAX >= INT64_MAX
+// Normal POSIX platforms are limited to SSIZE_MAX. However on 64-bit platforms there's no point
+// saturating any input I/O lengths - that would be 9 exabytes & insane to do in 1 I/O call.
+// Using SIZE_MAX ensures that the kj::min evaluation should be elided and the read/write/writev
+// calls should inlined directly to the libc function.
+#define KJ_OS_MAX_IO SIZE_MAX
+#else
+// On 32-bit platforms make sure that trying to do large I/O still works correctly (saturating &
```

Comment:
If I recall correctly (it's been a while since I first wrote & tested everything - should have taken better notes clearly), this was for FreeBSD which returns EINVAL when the value is > SSIZE_MAX. On 32-bit platforms means writes > 2GB would fail even though you would expect to be able to write 4GB. Does that sound reasonable? I think I wrote this comment *after* our discussion in the original bug report so I would have incorporated the feedback that oversized writes in general would be impossible...

https://www.freebsd.org/cgi/man.cgi?write(2)

Reply:
Hmm, is it actually feasible to allocate even a 2GB array on a 32-bit platform? I realize the virtual address space theoretically supports it but I would think there would be some practical limitations in the way, like kernel/user address space separation, fragmentation (especially with ASLR), etc.

Reply:
Coming back to this, apparently on MacOS, a 32-bit process can apparently allocate the full 4GB: https://forums.freebsd.org/threads/how-much-memory-freebsd-amd64-can-allocate-to-an-i386-binary.62171/page-2

That being said, macOS 10.15 released in fall 2019 doesn't support 32-bit apps anymore so what you wrote is probably true of all remaining systems. I don't know what minimum version of macOS is targeted.

Reply:
OK, so it's theoretically possible... but in practice, the application would really have to go out of its way to allocate one contiguous array over 2GB without being blocked by fragmentation, and it couldn't possibly allocate a second such array. Any app doing that should be aware that it is risking overflow and should be treading carefully. So I feel like this is not something we really need to contort our code to handle.

If we ignore that case, can the code be simpler? Can we avoid the need to override `write()` then? Or at least, can we put less stuff in the header? Maybe `KJ_OS_MAX_IO` could be kept private within `miniposix.c++`?